django/db/models/fields/related_lookups.py

```diff
@@ -105,7 +104,7 @@ def get_prep_lookup(self):
                     target_field = self.lhs.field.name
                 else:
                     target_field = self.lhs.field.target_field.name
-                self.rhs.add_fields([target_field], True)
+                self.rhs.set_values([target_field])
```

Comment:
Should we do the same in `In` and `Exact` lookups? i.e. use `set_values()` instead of
```python
self.rhs.clear_select_clause()
self.rhs.add_fields(["pk"])
```

Reply:
i dont think that would be required.

Reply:
Can I get a test case with ```Exact``` lookup ?

Reply:
Can i get the test cases for the same ?

Reply:
@felixxm have you confirmed it also affects `lookups.In`?

I agree that we should use `set_values` in `lookups.In.get_prep_lookup` as well for the sake of consistency but the changes in `RelatedIn.get_prep_lookup` where necessary to get tests passing because its call to `super()` fell back into this branch

https://github.com/django/django/blob/0c3981eb5094419fe200eb46c71b5376a2266166/django/db/models/lookups.py#L423-L425

As the manual calls to `clear_select_clause` + `add_fields` in `RelatedIn` didn't set `has_select_fields = True` and thus caused the simulated `values(target_field)` call to be overridden by `values("pk")` [causing the two reported tests failures](https://code.djangoproject.com/ticket/33975#comment:7).

All that to say that I agree with you that we should make these changes in `In.get_prep_lookup` but there are no ways to create a regession test for them since `In.get_prep_lookup` MRO doesn't rely on the presence of `has_select_fields = True` like `RelatedIn.get_prep_lookup` did.

Reply:
> have you confirmed it also affects `lookups.In`?

Yes. The suggested test:
```python
long_books_qs = (
    Book.objects.filter(
        pages__gt=400,
    )
    .annotate(book_annotate=Value(1))
    .alias(book_alias=Value(1))
)
Book.objects.filter(pk__in=long_books_qs)
```
crashes with `django.db.utils.OperationalError: sub-select returns 10 columns - expected 1` where:
```python
long_books_qs = (
    Book.objects.filter(
        pages__gt=400,
    )
    .annotate(book_annotate=Value(1))
    .annotate(book_alias=Value(1))
)
Book.objects.filter(pk__in=long_books_qs)
```
works properly.

> As the manual calls to `clear_select_clause` + `add_fields` in `RelatedIn` didn't set `has_select_fields = True` and thus caused the simulated `values(target_field)` call to be overridden by `values("pk")` [causing the two reported tests failures](https://code.djangoproject.com/ticket/33975#comment:7).
> 
> All that to say that I agree with you that we should make these changes in `In.get_prep_lookup` but there are no ways to create a regession test for them since `In.get_prep_lookup` MRO doesn't rely on the presence of `has_select_fields = True` like `RelatedIn.get_prep_lookup` did.

We probably need something more to fix `In` lookup :thinking:

Reply:
I think there's a misunderstanding here.

You test effectively fails against `main` but passes against this branch without any further changes. It also passes if we remove the changes in `RelatedIn.get_prep_lookup` that this thread is attached to.

The changes in `RelatedIn.get_prep_lookup` are necessary to make sure that pre-existing tests (the ones defined on `queries.tests.ToFieldTests`) are not broken by the changes to `has_select_fields` assignment in `sql.Query`.

With all that said, I agree that we should use `set_values` in `In.get_prep_lookup` but it's not something that is necessary to get any tests passing.

Reply:
I replaced 
```
self.rhs.clear_select_clause()
self.rhs.add_fields(["pk"])
```
with 
```
self.rhs.set_values(["pk"])
```
in ```In.get_prep_lookup``` and ```Exact.get_prep_lookup``` , still there was no failed testcase for ticket_33975 branch. Should I push those changes too?

Reply:
> You test effectively fails against main but passes against this branch without any further changes. It also passes if we remove the changes in RelatedIn.get_prep_lookup that this thread is attached to.

True, so it's a proper regression test for `In()` lookup, and at the same time using `set_values` is not necessary for this to work.

Reply:
> in In.get_prep_lookup and Exact.get_prep_lookup , still there was no failed testcase for this branch. Should I push those changes too?

Let's leave it for a separate clean-up.

Reply:
okay!